### ISSUE TITLE
2.1.1, 2.2.0 releases

### ISSUE THREAD

**Issue by tbekolay** [nengo/nengo#1080]
Making a separate issue to move the discussion out of #1010.

Proposed PRs to include in each release (feel free to suggest others in comments):

**2.1.1**

- [x] #1074
- [x] #1027
- [x] #1052
- [x] #1069 
- [x] #1088 
- [x] #1090 
- [x] #1091

**2.2.0**

- [ ] #1010
- [ ] #984
- [ ] #1045
- [ ] #1016
- [ ] #1107

---

One thing I'm not sure about are the PRs that make relatively large changes to SPA (e.g., #1006, #1014). I think @hunse and I aren't super well equipped to review them, and it doesn't seem quite right to up the minor version of Nengo whenever the SPA API changes (which seems to be happening relatively frequently). We sort of talked about it before, but I don't remember if there was a strong consensus, so I'll raise the issue again: is 2.2 the right time to move SPA to a different repo and Python package?

**Comment by hunse** [comment]
I'd also like #1069 (PR template), #1045 (ShapeParam), #1016 (multiple axis support for some utils), probably all fine in 2.2.0.

**Comment by hunse** [comment]
Also, I'm really in favour of moving SPA to a different repo, and I think 2.2 is a good time to do that since none of the PRs are touching SPA, so it shouldn't be too confusing.

**Comment by jgosmann** [comment]
I think I was in favour of having SPA in core Nengo so far? But I might change my mind on that. Having an external package would us allow to have a legacy SPA package that ensures full backwards compatibility, while we can create a new SPA API without worrying about backwards compatibility. However, there will be a few things where it is not clear whether to put then in core Nengo or the SPA package. That is mainly the spaopt optimization algorithm and the basic ensemble array for unit vectors (that does not need to be tied to any SPA stuff or vocabulary and might be useful for other things to).

**Comment by jgosmann** [comment]
List of PRs looks good. I think, #1069 does not need to be tied to version as it doesn't really affect the code base, but it helpful for all future PRs once it is in master. I prefer to have it merged sooner than later, but not before the remaining open points have been decided.

Another word on the potential Nengo/SPA split: As future SPA (mainly with spaopt) requires some Nengo features (deferrables or the equivalent), it would be nice to have them merged timely to simplify development. It would probably be an easier to develop SPA with just the Nengo master than having both packages on some development branch (that might both change).

**Comment by Seanny123** [comment]
I'm okay with these pull requests. As for SPA development, I think it's still a good idea to keep it in core until we finish the major changes (hierarchical, compiler, spaopt).

**Comment by jgosmann** [comment]
The progress bar widget is broken in the most recent Jupyter version. I would like to get a fix in for the 2.1.1 release. (I basically have a fix, but need to make it backwards compatible and prepare the PR.)

**Comment by jgosmann** [comment]
The PR for the progress bar fix is #1088.

**Comment by jgosmann** [comment]
I also want to bring fixing #1089 into the discussion. I'm undecided how important it is to fix this bug for 2.1.1 (without large vocabularies as parameters it is not as severe).

**Comment by tbekolay** [comment]
I added #1088, #1089 and #1090 to 2.1.1. Should all be relatively quick reviews.

**Comment by tbekolay** [comment]
OK, I think this issue has outlived its usefulness; tracking the 2.2.0 PRs in the [2.2.0 milestone](https://github.com/nengo/nengo/milestones/2.2.0%20release). If you think an issue / PR should be in / should not be in that milestone, please post a comment in that issue / PR.

**Comment by hunse** [comment]
There's a lot of stuff that was in the 2.2.0 milestone from before. I'm not terribly happy with our current system or lack thereof of deciding what should be in a particular milestone. For example, we often put something in a milestone and then later realize it's taking too long and move it out. I'd rather err on the other side, and keep things out of a milestone until they're pretty much ready to go. I think if we follow through with our plan to release often, people will be less concerned about trying to bundle their stuff into the next release.

Should we open an issue to discuss how to determine what goes in a milestone? Also, I'd be up for reducing the 2.2.0 milestone to pretty much the things we had had here, and then we can talk about adding more stuff. It would be really useful for me to have #1010 in particular in a release since there are things in `nengo_extras` that rely on that which I can't merge until we get that out.

**Comment by tbekolay** [comment]
Sure, I'm down for taking out everything but what we had in here. I'll open an issue to talk about planning out releases!
